tests/request_body_transient_error.rs

```diff
@@ -0,0 +1,385 @@
+        "temperature": 0.8
+    }"#
+    .as_bytes();
+
+    module
+        .call_proxy_on_request_body(http_context, body.len() as i32, true)
+        .expect_log(
+            Some(LogLevel::Debug),
+            Some("#2 on_http_request_body: body_size: 169, end_of_stream: true"),
+        )
+        .expect_get_buffer_bytes(Some(BufferType::HttpRequestBody))
+        .returning(Some(body))
+        .expect_log(
+            Some(LogLevel::Debug),
+            Some("#2 on_http_request_body (size: 169): action_set selected test-transient-errors")
+        )
+        // Expect the build_descriptor log when processing with complete body
+        .expect_log(
+            Some(LogLevel::Debug),
+            Some("build_descriptor(rl): starting, conditional_data_sets: 1"),
+        )
+        // Now the requestBodyJSON evaluation should succeed and grpc call should be made
+        .expect_log(Some(LogLevel::Debug), Some("#2 send_grpc_request: limitador-cluster envoy.service.ratelimit.v3.RateLimitService ShouldRateLimit 5s"))
+        .expect_grpc_call(
+            Some("limitador-cluster"),
+            Some("envoy.service.ratelimit.v3.RateLimitService"),
+            Some("ShouldRateLimit"),
+            Some(&[0, 0, 0, 0]),
+            None,
+            Some(5000),
+        )
+        .returning(Ok(42))
+        .execute_and_expect(ReturnType::Action(Action::Pause))
+        .unwrap();
+
+    // Step 4: gRPC response handling
+    let grpc_response: [u8; 2] = [8, 1]; // OK response
+    module
+        .call_proxy_on_grpc_receive(http_context, 42, grpc_response.len() as i32)
+        .expect_log(
+            Some(LogLevel::Debug),
+            Some("#2 on_grpc_call_response: received gRPC call response: token: 42, status: 0"),
+        )
+        .expect_get_buffer_bytes(Some(BufferType::GrpcReceiveBuffer))
+        .returning(Some(&grpc_response))
+        .expect_log(
+            Some(LogLevel::Debug),
+            Some("process_response(rl): received OK response"),
+        )
+        .execute_and_expect(ReturnType::None)
+        .unwrap();
+}
+
+#[test]
+#[serial]
+fn it_handles_multiple_transient_expressions_in_request_data() {
+    let args = tester::MockSettings {
+        wasm_path: wasm_module(),
+        quiet: false,
+        allow_unexpected: false,
+    };
+    let mut module = tester::mock(args).unwrap();
+
+    module
+        .call_start()
+        .execute_and_expect(ReturnType::None)
+        .unwrap();
+
+    let root_context = 1;
+    // Configuration with request_data that mixes transient (requestBodyJSON) and
+    // non-transient (request.method) expressions - this tests that the action waits
+    // for transient data while non-transient data can be evaluated immediately
+    let cfg = r#"{
+        "requestData": {
+            "metrics.labels.model": "requestBodyJSON('/model')",
+            "metrics.labels.max_tokens": "requestBodyJSON('/max_tokens')",
+            "metrics.labels.method": "request.method"
```

Comment:
This request Data expression `request.method` evaluates correctly because it's in the predicates section as well. If you remove `request-.method` from predicates, this should fail, as evaluating `request.method` at request body phase cannot be done. I will be fixing that in a separated PR